metrics_api/lib/opentelemetry/internal/proxy_meter.rb

```diff
@@ -26,31 +25,193 @@ def initialize
       #
       # @param [Meter] meter The Meter to delegate to
       def delegate=(meter)
-        @mutex.synchronize do
+        @delegate_mutex.synchronize do
           if @delegate.nil?
             @delegate = meter
-            @instrument_registry.each_value { |instrument| instrument.upgrade_with(meter) }
+
+            @instrument_registry.each_value do |proxy_instrument|
+              proxy_instrument.delegate =
+                case proxy_instrument
+                when ProxyInstrument::Counter
+                  meter.create_counter(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    advice:      proxy_instrument.advice
+                  )
+                when ProxyInstrument::Histogram
+                  meter.create_histogram(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    advice:      proxy_instrument.advice
+                  )
+                when ProxyInstrument::UpDownCounter
+                  meter.create_up_down_counter(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    advice:      proxy_instrument.advice
+                  )
+                when ProxyInstrument::ObservableCounter
+                  meter.create_observable_counter(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    callbacks:   proxy_instrument.callbacks
+                  )
+                when ProxyInstrument::ObservableGauge
+                  meter.create_observable_gauge(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    callbacks:   proxy_instrument.callbacks
+                  )
+                when ProxyInstrument::ObservableUpDownCounter
+                  meter.create_observable_up_down_counter(
+                    proxy_instrument.name,
+                    unit:        proxy_instrument.unit,
+                    description: proxy_instrument.description,
+                    callbacks:   proxy_instrument.callbacks
+                  )
+                end
+            end
```

Comment:
This is more verbose now because we have a Proxy Instrument class for each Instrument